ee/tabby-db/src/github_oauth_credential.rs

```diff
@@ -0,0 +1,107 @@
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use rusqlite::{params, OptionalExtension};
+
+use super::DbConn;
+
+const GITHUB_OAUTH_CREDENTIAL_ROW_ID: i32 = 1;
+
+pub struct GithubOAuthCredentialDAO {
+    pub client_id: String,
+    pub client_secret: String,
+    pub active: bool,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+}
+
+impl GithubOAuthCredentialDAO {
+    fn from_row(row: &rusqlite::Row<'_>) -> std::result::Result<Self, rusqlite::Error> {
+        Ok(Self {
+            client_id: row.get(0)?,
+            client_secret: row.get(1)?,
+            active: row.get(2)?,
+            created_at: row.get(3)?,
+            updated_at: row.get(4)?,
+        })
+    }
+}
+
+/// db read/write operations for `github_oauth_credential` table
+impl DbConn {
+    pub async fn update_github_oauth_credential(
+        &self,
+        client_id: &str,
+        client_secret: &str,
+        active: bool,
+    ) -> Result<()> {
+        let client_id = client_id.to_string();
+        let client_secret = client_secret.to_string();
+
+        self.conn
+            .call(move |c| {
+                let mut stmt = c.prepare(
+                    r#"INSERT INTO github_oauth_credential (id, client_id, client_secret)
+                    VALUES (?, ?, ?) ON CONFLICT(id) DO UPDATE
+                    SET client_id = ?, client_secret = ?, active = ?, updated_at = datetime('now')
+                    WHERE id = ?"#,
+                )?;
+                stmt.insert(params![
```

Comment:
you could use positioned parameter or named parameter to avoid passing args multiple times

https://docs.rs/rusqlite/latest/rusqlite/trait.Params.html#example-positional

Reply:
This is really helpful! 👍